In [1]:
import pandas as pd

### Train

In [2]:
compressed_data = pd.read_csv('../../Dataset/train/A_2017.11.24_14.34.43.csv.gz', compression='gzip')

In [3]:
compressed_data.head(10)

Timestamp  Longitude   Latitude  Speed Operatorname  CellID  \
0  2017.11.24_14.34.43  -8.458313  51.901641      0            A       1   
1  2017.11.24_14.34.43  -8.458313  51.901641      0            A       1   
2  2017.11.24_14.34.44  -8.458360  51.901481      2            A       1   
3  2017.11.24_14.34.45  -8.458360  51.901481      2            A       1   
4  2017.11.24_14.34.46  -8.458360  51.901481      2            A       1   
5  2017.11.24_14.34.47  -8.458360  51.901481      2            A       1   
6  2017.11.24_14.34.48  -8.458360  51.901481      2            A       1   
7  2017.11.24_14.34.49  -8.458360  51.901481      2            A       1   
8  2017.11.24_14.34.50  -8.458360  51.901481      2            A       1   
9  2017.11.24_14.34.51  -8.458360  51.901481      2            A       1   

  NetworkMode  RSRP RSRQ SNR CQI RSSI  DL_bitrate  UL_bitrate State NRxRSRP  \
0         LTE   -88  -12  17  10  -71          49          36     D       -   
1         LTE   -88  -12  17  10  -71          49          36     D       -   
2         LTE   -88  -12   8   9  -73          49          36     D       -   
3         LTE   -88  -12  17  10  -71         173           6     D       -   
4         LTE   -85  -12  16  10  -69           0           0     I       -   
5         LTE   -85  -12  16  10  -75           0           0     I       -   
6         LTE   -84  -12  19  11  -74           0           0     I       -   
7         LTE   -84  -12  19  12  -68           0           0     I       -   
8         LTE   -85  -12  16  12  -72          39          25     D       -   
9         LTE   -85  -12  16  11  -72           0           0     D       -   

  NRxRSRQ ServingCell_Lon ServingCell_Lat ServingCell_Distance  
0       -       -8.460674        51.89841   394.10000000000002  
1       -       -8.460674        51.89841   394.10000000000002  
2       -       -8.460674        51.89841   376.57999999999998  
3       -       -8.460674        51.89841   376.57999999999998  
4       -       -8.460674        51.89841   376.57999999999998  
5       -       -8.460674        51.89841   376.57999999999998  
6       -       -8.460674        51.89841   376.57999999999998  
7       -       -8.460674        51.89841   376.57999999999998  
8       -       -8.460674        51.89841   376.57999999999998  
9       -       -8.460674        51.89841   376.57999999999998

In [4]:
compressed_data.tail(10)

Timestamp  Longitude   Latitude  Speed Operatorname  CellID  \
9441  2017.11.24_17.14.02  -6.296212  53.346171      2            A     896   
9442  2017.11.24_17.14.02  -6.296212  53.346171      2            A     896   
9443  2017.11.24_17.14.03  -6.296212  53.346171      2            A     896   
9444  2017.11.24_17.14.05  -6.296212  53.346171      2            A     896   
9445  2017.11.24_17.14.06  -6.296212  53.346171      2            A     896   
9446  2017.11.24_17.14.07  -6.296212  53.346171      2            A     896   
9447  2017.11.24_17.14.08  -6.296212  53.346171      2            A     896   
9448  2017.11.24_17.14.09  -6.296212  53.346171      2            A     896   
9449  2017.11.24_17.14.10  -6.296212  53.346171      2            A     896   
9450  2017.11.24_17.14.11  -6.296212  53.346171      2            A     896   

     NetworkMode  RSRP RSRQ SNR CQI RSSI  DL_bitrate  UL_bitrate State  \
9441       HSPA+   -85   -2   -   -    -           0           2     D   
9442       HSPA+   -85   -2   -   -    -           2           2     D   
9443       HSPA+   -85   -2   -   -    -           0           0     I   
9444       HSPA+   -83   -2   -   -    -           0           0     I   
9445       HSPA+   -83   -2   -   -    -           0           0     I   
9446       HSPA+   -85   -2   -   -    -           0           0     I   
9447       HSPA+   -85   -2   -   -    -           0           0     I   
9448       HSPA+   -85   -2   -   -    -           0           0     I   
9449       HSPA+   -85   -2   -   -    -           0           0     I   
9450       HSPA+   -85   -2   -   -    -           0           0     I   

     NRxRSRP NRxRSRQ ServingCell_Lon ServingCell_Lat ServingCell_Distance  
9441   -51.0   -24.0               -               -                    -  
9442   -51.0   -24.0               -               -                    -  
9443   -51.0   -24.0               -               -                    -  
9444   -51.0   -24.0               -               -                    -  
9445   -51.0   -24.0               -               -                    -  
9446   -51.0   -24.0               -               -                    -  
9447   -51.0   -24.0               -               -                    -  
9448   -51.0   -24.0               -               -                    -  
9449   -51.0   -24.0               -               -                    -  
9450   -51.0   -24.0               -               -                    -

In [5]:
16 + 53 + 31 + 15 + 20

135

In [6]:
compressed_data.shape

(9451, 20)

In [7]:
60 * 15

900

In [8]:
120 * 60 

7200

In [9]:
40 * 60 

2400

In [10]:
7200 + 2400 + 54

9654

Deductively, the data was collected in a one-second interval. If that's the case, then this particular data was collected from time: `2:34:43` to `5:14:11` : Which is `2 hours, 40 minutes and 54s` rounding up to about `9654 seconds`. This is abit close to the number of interval/traces/rows from the data having `9451` recorded values within these periods. 

But, if the data or trace takes ~2hrs to be recorded, what do they mean by each trace lasting approximately `15 minutes`?

### confirmation

In [11]:
compressed_data.columns

Index(['Timestamp', 'Longitude', 'Latitude', 'Speed', 'Operatorname', 'CellID',
       'NetworkMode', 'RSRP', 'RSRQ', 'SNR', 'CQI', 'RSSI', 'DL_bitrate',
       'UL_bitrate', 'State', 'NRxRSRP', 'NRxRSRQ', 'ServingCell_Lon',
       'ServingCell_Lat', 'ServingCell_Distance'],
      dtype='object')

In [12]:
# Convert 'Timestamp' column to datetime
compressed_data['Timestamp'] = pd.to_datetime(compressed_data['Timestamp'], format='%Y.%m.%d_%H.%M.%S')

# Calculate the duration
start_time = compressed_data['Timestamp'].min()
end_time = compressed_data['Timestamp'].max()
duration = end_time - start_time

print(f"Start Time: {start_time}")
print(f"End Time: {end_time}")
print(f"Duration: {duration}")

# Check the number of rows to ensure 1-second intervals
expected_rows = duration.total_seconds() + 1  # Adding 1 to include both start and end seconds
actual_rows = len(compressed_data)
print(f"Expected Rows: {expected_rows}")
print(f"Actual Rows: {actual_rows}")

Start Time: 2017-11-24 14:34:43
End Time: 2017-11-24 17:14:11
Duration: 0 days 02:39:28
Expected Rows: 9569.0
Actual Rows: 9451


In [13]:
hour_sec = 60 * 2 * 60
mins_sec = 39 * 60
sec = 28
total = hour_sec + mins_sec + sec
print(f'2 hours to seconds: {hour_sec}')
print(f'39 minutes to seconds: {mins_sec}')

print(f'\n Total trace time/rows: {total}')

2 hours to seconds: 7200
39 minutes to seconds: 2340

 Total trace time/rows: 9568


In [14]:
compressed_data.columns

Index(['Timestamp', 'Longitude', 'Latitude', 'Speed', 'Operatorname', 'CellID',
       'NetworkMode', 'RSRP', 'RSRQ', 'SNR', 'CQI', 'RSSI', 'DL_bitrate',
       'UL_bitrate', 'State', 'NRxRSRP', 'NRxRSRQ', 'ServingCell_Lon',
       'ServingCell_Lat', 'ServingCell_Distance'],
      dtype='object')

In [15]:
compressed_data.head()

Timestamp  Longitude   Latitude  Speed Operatorname  CellID  \
0 2017-11-24 14:34:43  -8.458313  51.901641      0            A       1   
1 2017-11-24 14:34:43  -8.458313  51.901641      0            A       1   
2 2017-11-24 14:34:44  -8.458360  51.901481      2            A       1   
3 2017-11-24 14:34:45  -8.458360  51.901481      2            A       1   
4 2017-11-24 14:34:46  -8.458360  51.901481      2            A       1   

  NetworkMode  RSRP RSRQ SNR CQI RSSI  DL_bitrate  UL_bitrate State NRxRSRP  \
0         LTE   -88  -12  17  10  -71          49          36     D       -   
1         LTE   -88  -12  17  10  -71          49          36     D       -   
2         LTE   -88  -12   8   9  -73          49          36     D       -   
3         LTE   -88  -12  17  10  -71         173           6     D       -   
4         LTE   -85  -12  16  10  -69           0           0     I       -   

  NRxRSRQ ServingCell_Lon ServingCell_Lat ServingCell_Distance  
0       -       -8.460674        51.89841   394.10000000000002  
1       -       -8.460674        51.89841   394.10000000000002  
2       -       -8.460674        51.89841   376.57999999999998  
3       -       -8.460674        51.89841   376.57999999999998  
4       -       -8.460674        51.89841   376.57999999999998

In [16]:
compressed_data.NetworkMode.unique()

array(['LTE', 'HSPA+', 'UMTS', 'EDGE', 'HSUPA'], dtype=object)

NOTES:
* Understand GPS cordinates with longiture and latitude
* 

### Bus

In [17]:
bus_data = pd.read_csv('../../Dataset/bus/A_2017.11.30_16.48.26.csv.gz', compression='gzip')

In [18]:
bus_data.head()

Timestamp  Longitude   Latitude  Speed Operatorname  CellID  \
0  2017.11.30_16.48.26  -8.501373  51.893359      0            A       2   
1  2017.11.30_16.48.26  -8.501291  51.893462      1            A       2   
2  2017.11.30_16.48.27  -8.501291  51.893462      1            A       2   
3  2017.11.30_16.48.28  -8.501291  51.893462      1            A       2   
4  2017.11.30_16.48.29  -8.501291  51.893462      1            A       2   

  NetworkMode  RSRP  RSRQ   SNR  CQI RSSI  DL_bitrate  UL_bitrate State  \
0         LTE  -102   -12  10.0    7  -85           3           7     D   
1         LTE  -102   -12  10.0    7  -85           3           7     D   
2         LTE  -102   -12   7.0   10  -87         310          14     D   
3         LTE  -102   -12   7.0    7  -85           0           0     I   
4         LTE  -102   -13   8.0    7  -85           0           0     I   

  NRxRSRP NRxRSRQ ServingCell_Lon ServingCell_Lat ServingCell_Distance  
0       -       -       -8.491719       51.893905   665.24000000000001  
1       -       -       -8.491719       51.893905   658.67999999999995  
2       -       -       -8.491719       51.893905   658.67999999999995  
3       -       -       -8.491719       51.893905   658.67999999999995  
4       -       -       -8.491719       51.893905   658.67999999999995

In [19]:
bus_data.tail()

Timestamp  Longitude   Latitude  Speed Operatorname  CellID  \
905  2017.11.30_17.04.24  -8.556712  51.892251     46            A       1   
906  2017.11.30_17.04.25  -8.556712  51.892251     46            A       1   
907  2017.11.30_17.04.26  -8.556712  51.892251     46            A       1   
908  2017.11.30_17.04.27  -8.556712  51.892251     46            A       1   
909  2017.11.30_17.04.27  -8.556712  51.892251     46            A       1   

    NetworkMode  RSRP  RSRQ  SNR  CQI RSSI  DL_bitrate  UL_bitrate State  \
905         LTE   -91   -12  3.0   12  -72        7068         128     D   
906         LTE   -91   -12  3.0   10  -69        8992         166     D   
907         LTE   -86   -11  5.0    9  -76        9584         177     D   
908         LTE   -86   -11  5.0   10  -74       11060         198     D   
909         LTE   -86   -11  5.0    5  -76       11060         198     D   

    NRxRSRP NRxRSRQ ServingCell_Lon ServingCell_Lat ServingCell_Distance  
905   -89.0   -10.0       -8.535593       51.880268   1968.8299999999999  
906   -89.0   -10.0       -8.535593       51.880268   1968.8299999999999  
907   -85.0    -9.0       -8.535593       51.880268   1968.8299999999999  
908   -85.0    -9.0       -8.535593       51.880268   1968.8299999999999  
909   -85.0    -9.0       -8.535593       51.880268   1968.8299999999999

In [20]:
bus_data.columns

Index(['Timestamp', 'Longitude', 'Latitude', 'Speed', 'Operatorname', 'CellID',
       'NetworkMode', 'RSRP', 'RSRQ', 'SNR', 'CQI', 'RSSI', 'DL_bitrate',
       'UL_bitrate', 'State', 'NRxRSRP', 'NRxRSRQ', 'ServingCell_Lon',
       'ServingCell_Lat', 'ServingCell_Distance'],
      dtype='object')

In [21]:
bus_data.Operatorname.nunique()

1

In [22]:
bus_2_data = pd.read_csv('../../Dataset/bus/A_2018.01.25_16.33.53.csv.gz', compression='gzip')

In [23]:
bus_2_data.Operatorname.nunique()

1

In [24]:
bus_2_data.head()

Timestamp  Longitude   Latitude  Speed Operatorname  CellID  \
0  2018.01.25_16.33.53  -8.499627  51.893860      0            A       2   
1  2018.01.25_16.33.54  -8.498502  51.893994     34            A       2   
2  2018.01.25_16.33.54  -8.498502  51.893994     34            A       2   
3  2018.01.25_16.33.55  -8.498502  51.893994     34            A       2   
4  2018.01.25_16.33.56  -8.498502  51.893994     34            A       2   

  NetworkMode  RSRP  RSRQ  SNR CQI RSSI  DL_bitrate  UL_bitrate State NRxRSRP  \
0         LTE   -88   -13  4.0   7  -70        1476          28     D       -   
1         LTE   -88   -13  6.0   7  -74        1476          28     D       -   
2         LTE   -91   -14  6.0   7  -74        1610          30     D       -   
3         LTE   -91   -14  6.0   7  -80        2102          70     D       -   
4         LTE   -96   -14  5.0   7  -80        1554          54     D       -   

  NRxRSRQ ServingCell_Lon ServingCell_Lat ServingCell_Distance  
0       -               -               -                    -  
1       -               -               -                    -  
2       -               -               -                    -  
3       -               -               -                    -  
4       -               -               -                    -

In [25]:
bus_3_data = pd.read_csv('../../Dataset/bus/A_2018.01.25_17.27.30.csv.gz', compression='gzip')

In [28]:
bus_3_data.Operatorname.unique()

array(['A', '27205', '0'], dtype=object)

In [27]:
bus_3_data.head()

Timestamp  Longitude   Latitude  Speed Operatorname  CellID  \
0  2018.01.25_17.27.30  -8.473233  51.898205      2            A       1   
1  2018.01.25_17.27.30  -8.473233  51.898205      2            A       1   
2  2018.01.25_17.27.31  -8.473518  51.898446      0            A       1   
3  2018.01.25_17.27.31  -8.473518  51.898446      0            A       1   
4  2018.01.25_17.27.32  -8.473518  51.898446      0            A       1   

  NetworkMode  RSRP  RSRQ  SNR CQI RSSI  DL_bitrate  UL_bitrate State NRxRSRP  \
0         LTE   -95   -13  1.0   7  -81           0           0     D       -   
1         LTE   -95   -13  5.0   7  -81          11           0     D       -   
2         LTE   -95   -13  5.0   7  -81          11           0     D       -   
3         LTE   -95   -13  4.0   8  -78           0           0     I       -   
4         LTE   -95   -13  4.0   8  -78        4988         187     D       -   

  NRxRSRQ ServingCell_Lon ServingCell_Lat ServingCell_Distance  
0       -       -8.472275       51.895123   348.94999999999999  
1       -       -8.472275       51.895123   348.94999999999999  
2       -       -8.472275       51.895123   379.22000000000003  
3       -       -8.472275       51.895123   379.22000000000003  
4       -       -8.472275       51.895123   379.22000000000003

In [37]:
random_bus_data = pd.read_csv('../../Dataset/bus/B_2018.01.27_12.09.59.csv.gz', compression='gzip')

In [38]:
random_bus_data.Operatorname.unique()

array(['B'], dtype=object)

In [39]:
random_bus_data.NetworkMode.unique()

array(['LTE'], dtype=object)